src/Algebra/Graph/Algorithm/Internal.hs

```diff
@@ -1,30 +1,38 @@
 module Algebra.Graph.Algorithm.Internal where
 
-import           Algebra.Graph
+import                            Algebra.Graph
+import Data.DisjointSet           (DisjointSet)
 import qualified Data.DisjointSet as DS
-import qualified Data.Set         as S
 
 -- | Extract disjoint set of connectivity components from the graph
-components :: Ord a => Graph a -> DS.DisjointSet a
+--
+-- >>> DS.pretty $ components ((1 * 2) + (3 * 4))
+-- "{{1,2},{3,4}}"
+--
+-- >>> DS.pretty $ components (((1 * 2) + (3 * 4)) + (2 * 3))
+-- "{{1,2,3,4}}"
+components :: Ord a => Graph a -> DisjointSet a
 components Empty         = DS.empty
 components (Vertex x)    = DS.singleton x
-components (Overlay x y) = mergeComponents (components x) (components y)
-components (Connect x y) = DS.fromLists [vertexList x ++ vertexList y]
+components (Connect x y) = DS.fromLists [extractVertices x <> extractVertices y]
+components (Overlay x y) = (components x) <> (components y)
```

Comment:
```suggestion
components (Overlay x y) = components x <> components y
```

This relies on `DS.append` (which is not exported for some reason) which seems to be O(m log n), m <= n.
So overall complexity of `components` right now seems to be O(n log^2 n).

I think it should be possible to provide a more efficient version of `DS.append` similar to `Map.union`, which should eliminate one extra _log n_.